### PR TITLE
operator: allow updating pod template in the sts

### PR DESCRIPTION
**Pull request by vanyarock01**
### Problem
There is currently no way to customize the pod template spec.

### Motivation
Without this, it is impossible to set such important parameters as:
- node selectors
- traits
- tolerations
- and others

### Solution
Now the operator likes to control everything. The settings that are updated in the CRD "replicasetTemplate" operator selectively transfer to the statefulset spec.

1. Let's completely change the pod template
2. Allowed to change labels and annotations (except "internal")

Internal it:
* StatefulSet labels - `tarantool.io/replicaset-uuid`, `tarantool.io/vshardGroupName`
* StatefulSet annotations - `tarantool.io/isBootstrapped`, `tarantool.io/replicaset-weight`,
* StatefulSet.Spec.Template labels - `tarantool.io/replicaset-uuid`, `tarantool.io/vshardGroupName`

### CODE REVIEW

File: pkg/controller/role/role_controller.go [hunk 1]
```diff
@@ -290,18 +295,10 @@ func CreateStatefulSetFromTemplate(replicasetNumber int, name string, role *tara
 
 	sts.Spec.UpdateStrategy = appsv1.StatefulSetUpdateStrategy{Type: "OnDelete"}
```
Comment (R-omk) [inline pkg/controller/role/role_controller.go:296]: This also needs to be able to be set in the template

File: pkg/controller/role/role_controller.go [hunk 2]
```diff
@@ -227,48 +223,59 @@ func (r *ReconcileRole) Reconcile(request reconcile.Request) (reconcile.Result,
...
+		}
 
-				sts.ObjectMeta.Annotations["tarantool.io/rolesToAssign"] = templateRolesToAssign
-				sts.Spec.Template.Annotations["tarantool.io/rolesToAssign"] = templateRolesToAssign
+		// immutable container template labels
+		for _, key := range []string{
+			"tarantool.io/replicaset-uuid",
+			"tarantool.io/vshardGroupName",
+		} {
+			template.Spec.Template.ObjectMeta.Labels[key] = sts.Spec.Template.ObjectMeta.Labels[key]
```
Comment (to6ka) [inline pkg/controller/role/role_controller.go:263]: * no `r.client.Update` for this case
* expected update both Labels with Annotations

File: pkg/controller/role/role_controller.go [hunk 3]
```diff
@@ -227,48 +223,59 @@ func (r *ReconcileRole) Reconcile(request reconcile.Request) (reconcile.Result,
...
+			sts.Spec.Template.Spec = template.Spec.Template.Spec
+			if err := r.client.Update(context.TODO(), &sts); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+
+		if template.Spec.UpdateStrategy.Type != sts.Spec.UpdateStrategy.Type {
+			reqLogger.Info("Updating sts UpdateStrategy", "new type", template.Spec.UpdateStrategy)
+			sts.Spec.UpdateStrategy = template.Spec.UpdateStrategy
 			if err := r.client.Update(context.TODO(), &sts); err != nil {
```
Comment (to6ka) [inline pkg/controller/role/role_controller.go:237]: you can make just one update at the end

File: pkg/controller/role/role_controller.go [hunk 4]
```diff
@@ -227,48 +223,59 @@ func (r *ReconcileRole) Reconcile(request reconcile.Request) (reconcile.Result,
 			}
```
Comment (to6ka) [inline pkg/controller/role/role_controller.go:223]: `template.Spec.Replicas != sts.Spec.Replicas` - pointer comparison

File: pkg/controller/role/role_controller.go [hunk 5]
```diff
@@ -227,48 +223,59 @@ func (r *ReconcileRole) Reconcile(request reconcile.Request) (reconcile.Result,
...
 		}
```
Comment (to6ka) [inline pkg/controller/role/role_controller.go:224]: false positive updates by operator
```
{"level":"info","ts":1629304303.5663705,"logger":"controller_role","msg":"Updating replicas count","Request.Namespace":"tarantool","Request.Name":"cld-recheck-router"}
{"level":"info","ts":1629304303.6102424,"logger":"controller_role","msg":"Updating template spec","Request.Namespace":"tarantool","Request.Name":"cld-recheck-router"}
{"level":"info","ts":1629304303.6559029,"logger":"controller_role","msg":"Updating sts Annotations","Request.Namespace":"tarantool","Request.Name":"cld-recheck-router"}
```

**Comment by vanyarock01** [comment]
Closed by https://github.com/tarantool/tarantool-operator/issues/83

**Comment by kluevandrew** [comment]
Will be implemented as part of a conceptually new version of the operator in the near future.
Rejected due to lack of plans to develop the current operator code.

**Pull request closed without merging**